docs/rfcs/component-universal-telemetry.md

```diff
@@ -0,0 +1,158 @@
+are unified by the component implementation. For example, the `otlp` receiver is a singleton, so its telemetry is not specific to a signal.
+Similarly, the `memory_limiter` processor is a singleton, so its telemetry is not specific to a pipeline.
+
+## Auto-Instrumentation Mechanism
+
+The mechanism of telemetry capture should be _external_ to components. Specifically, we should observe telemetry at each point where a
+component passes data to another component, and, at each point where a component consumes data from another component. In terms of the
+component graph, every _edge_ in the graph will have two layers of instrumentation - one for the producing component and one for the
+consuming component. Importantly, each layer generates telemetry ascribed to a single component instance, so by having two layers per
+edge we can describe both sides of each handoff independently.
+
+Telemetry captured by this mechanism should be associated with an instrumentation scope corresponding to the package which implements
+the mechanism. Currently, that package is `service/internal/graph`, but this may change in the future. Notably, this telemetry is not
+ascribed to individual component packages, both because the instrumentation scope is intended to describe the origin of the telemetry,
+and because no mechanism is presently identified which would allow us to determine the characteristics of a component-specific scope.
+
+### Auto-Instrumented Metrics
+
+There are two straightforward measurements that can be made on any pdata:
+
+1. A count of "items" (spans, data points, or log records). These are low cost but broadly useful, so they should be enabled by default.
+2. A measure of size, based on [ProtoMarshaler.Sizer()](https://github.com/open-telemetry/opentelemetry-collector/blob/9907ba50df0d5853c34d2962cf21da42e15a560d/pdata/ptrace/pb.go#L11).
+  These are high cost to compute, so by default they should be disabled (and not calculated).
+
+The location of these measurements can be described in terms of whether the data is "consumed" or "produced", from the perspective of the
+component to which the telemetry is ascribed.
+
+1. Incoming measurements are attributed to the component which is _consuming_ the data.
+2. Outgoing measurements are attributed to the component which is _producing_ the data.
+
+For both metrics, an `outcome` attribute with possible values `success` and `failure` should be automatically recorded, corresponding to
+whether or not the corresponding function call returned an error. Specifically, consumed measurements will be recorded with `outcome` as
+`failure` when a call from the previous component the `ConsumeX` function returns an error, and `success` otherwise. Likewise, produced
+measurements will be recorded with `outcome` as `failure` when a call to the next consumer's `ConsumeX` function returns an error, and
+`success` otherwise.
+
+```yaml
+    otelcol_component_consumed_items:
```

Comment:
Minor correction - receivers would not report a "consumed" metric based on my proposal, since this cannot be automatically captured (as the mechanism of consuming varies for each receiver).

If I'm understanding your point though, I think it is essentially the same as made by Bogdan [here](https://github.com/open-telemetry/opentelemetry-collector/pull/11406#discussion_r1803413871).

> I think it would essentially be useful to document some of the equations put forth in [this OTEP](https://github.com/open-telemetry/oteps/pull/259/files#diff-8ab369bae50cfda348c0bd20882db8a1bdbbffb979abc70bff298192a948588bR245).

If you have specific suggestions please share them.